gfp.go

```diff
@@ -16,6 +20,30 @@ func newGFp(x int64) (out *gfP) {
 	return out
 }
 
+func os2ip(b [32]byte) [4]uint64 {
+	t := [4]uint64{}
+	for w := 0; w < 4; w++ {
+		t[w] = binary.LittleEndian.Uint64(b[8*w : 8*w+8])
+	}
+	return t
+}
+
+// hashToBase implements hashing a message to an element of the field.
+// It follows the recommendations from https://tools.ietf.org/pdf/draft-irtf-cfrg-hash-to-curve-04.pdf
+// Note that:
+//      - we don't use HKDF-Extract and HKDF-Expand procedures as the field is prime,
+//      - the bias introduced by interpreting a 256-bit hash as an integer modulo p is negligeble.
+func hashToBase(msg []byte) *gfP {
```

Comment:
It would be great to have a method that is complaint with the draft.

Reply:
I tried to make it so. Can you point what is missing? I noted in the comment why don't use HDKF and that while the distribution of the hashes may not be perfectly uniform, the bias is negligible.

Reply:
It is true that the bias is negligible for this prime. However, I would be desirable to maintain compatibility with the draft, i.e. implementing hash_to_base following the recipe in the draft.
Doing so, it allows to compare test vectors between different implementations.

Reply:
To make the implementation compatible with the draft I would need to:
1) follow the instructions from Section 5.1 
"The   hash_to_curve function takes such a tag as a parameter, DST; this is   the recommended way of applying domain separation.  As an   alternative, implementations MAY instead prepend a domain separation   tag to the input msg; in this case, DST SHOULD be the empty string."
I suppose, you don't want to add such a parameter to HashG1 method, so I'd need to change dstMsg to be ""
2) use HKDF which for prime field would be just unnecessary computation as the loop in the draft is over 1...m, where m=1 in our case. 
For me it doesn't seem worth it, but if you insist, then is it ok to use golang.org/x/crypto/hkdf lib?

Reply:
>    I suppose, you don't want to add such a parameter to HashG1 method, so I'd need to change dstMsg to be ""

That is right.

> For me it doesn't seem worth it, but if you insist, then is it ok to use golang.org/x/crypto/hkdf lib?

That library is ok.

Reply:
I've implemented the changes but it introduces quite big overhead due to extra mod p operation and hdkf.

Old
BenchmarkHashG1Size8bytes-4   	    5000	    255832 ns/op	   0.03 MB/s
BenchmarkHashG1Size1k-4       	    5000	    361977 ns/op	   2.83 MB/s
BenchmarkHashG1Size8k-4       	    3000	    409785 ns/op	  19.99 MB/s
New
BenchmarkHashG1Size8bytes-4   	    3000	    370105 ns/op	   0.02 MB/s
BenchmarkHashG1Size1k-4       	    5000	    303564 ns/op	   3.37 MB/s
BenchmarkHashG1Size8k-4       	    2000	    514686 ns/op	  15.92 MB/s

Reply:
@armfazh ?

Reply:
Sorry for the delay, I will review your changes during the weekend.

Reply:
@armfazh did you have time to take a look at the PR?

Reply:
>    I suppose, you don't want to add such a parameter to HashG1 method, so I'd need to change dstMsg to be ""

As I mentioned before, the draft is constantly changing. One of the latest changes requires the `dst` (domain separation string) parameter.
See: https://github.com/cfrg / draft-irtf-cfrg-hash-to-curve/pull/170

Reply:
added